### ISSUE TITLE
add platt scaling / isotonic regression

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr#1121]
one should be able to put this on regression learners. 
@larskotthoff  did this PR #549 

and we need to figure out how to put this on classifiers with "uncalibrated" scores

**Comment by rwarnung** [comment]
There is code for this in caret posted [here](https://www.analyticsvidhya.com/blog/2016/07/platt-scaling-isotonic-regression-minimize-logloss-error/).

In fact for platt scaling you put in a meta model where the task is to calibrate a logistic regression where the target is the original target and the data are the "probabilities" (often only scores) of you primary learner - pred$data$prob.pos.

I think this is quite standard in scikit-learn and also described [here](http://scikit-learn.org/stable/modules/calibration.html).

I wrote code for myself in caret to do this. It takes the train set, returns the platt-scaling (a glm) and returns the scaled probabilities for a new set of probabilies.

getPlattScaling<-function(probtrain,testingtarget,probnew){
  meta.data = data.frame(prob=probtrain,target=testingtarget)
  platt.model = train(target~prob,data = meta.data,
                      method="glm",family=binomial, 
                      trControl =  trainControl(method="none"), metric = "logLoss")
  return(list(pred=predict(platt.model,newdata=data.frame(prob=probnew),type="prob"),platt.model=platt.model))
}

**Comment by stevenpawley** [comment]
Picking up on this conversation, would this be the same as making a stackedLearner from the original classifier which requires probability calibration, and with the super.learner being a classifier that produces well-calibrated probabilities, e.g. glm for platt rescaling? If the method was equal to 'stack.cv' then this appears similar to the approach using in sklearn's CalibratedClassifierCV?

**Comment by rwarnung** [comment]
Thank you for picking up. I agree at seveal points:

1. the sklearn link that you provide states exactly the two approach: Platt and isotonic

2. And another yes: one could see this as small ensemble. Just one base learner (the original one) and a glm on top (thus stacking). In order not to do pure overfitting I would suggest to do the fitting of the second step (the stacking) on an unseen sample - which is data demanding. I wonder: is it worth the trouble? Do we have references that e.g. Platt scaling improves classifiers? If yes, then under which circumstances?

**Comment by SteveBronder** [comment]
Would this be better to have in mlrcpo?

**Comment by rwarnung** [comment]
@SteveBronder the functionalities for ensembles is already in mlr. mlrcpo is for preprocessing (only, as far as I know) and this would be some kind of postprocessing. Or os mlrcpo meant for postprocessing too :) Both could be good places.

**Comment by SteveBronder** [comment]
Yes from the below I think you would use `makeCPOExtendedTargetOp()`

https://cran.r-project.org/web/packages/mlrCPO/vignettes/z_4_custom_CPOs_terse.html#target-operation-cpos

**Comment by SteveBronder** [comment]
Pinging @mb706

**Comment by rwarnung** [comment]
@SteveBronder I don't have any experience with `makeCPOExtendedTargetOp()` but it looks as if it would fit for the purpose. I don't know how it is done in sklearn but I wonder whether it leads to overfitting if I apply the `glm `on exactly the same data as where the primal learner was trained on. Thank you!

**Comment by mb706** [comment]
I don't think this can be implemented using CPO in its current form, and in any way it would be stretching the "data piping" metaphor quite a bit, so I'm not sure if CPO *should* be able to do this. The problem is that the CPO does not see the output of the model being trained, it only sees the incoming data. The justification for this is that a CPO should be "useful" even without having a model at hand: `data %>>% cpo` should work similarly well to `cpo %>>% learner`.

It would therefore *not* be possible to have something like `train(cpoPlattScale() %>>% learner, data)`, since the glm would need to have access to the prediction made by the Learner.

Instead of attaching the learner at the end of the pipeline, another idea would be to have a `CPO` that uses the Learner *internally* and converts a Task's target column to the predictions made by a model. It would look something like
```R
train(cpoTargetToPredictedProb(learner) %>>%
        # renames "target" to "oldtarget" column and creates
        # new target column with "prob" predictions by learner;
        # drops the other columns.
    makeLearner("classif.logreg"),
  data)
```
The problem with this approach is that target operating CPOs are not allowed to change non-target columns, and would therefore not be allowed to create the "oldtarget" column (and drop all the other columns). There *might* be a way to chain a feature CPO and a target CPO to accomplish this, but it would not be pretty. I could modify CPO with some work to allow this kind of construct, but I don't think it would add much (although I am open to be convinced otherwise).

To see what CPO *can* do, look at [cpoRegrResiduals](https://github.com/mlr-org/mlrCPO/blob/master/R/CPO_regrResiduals.R), which transforms a Task's target column into a column of (resampling) residuals of a primary learner, lets a secondary learner do prediction on these residuals, and then incorporates the result with the result of the primary learner for new data.

**Comment by Prometheus77** [comment]
Platt scaling/isotonic regression are most useful when combining base
classifiers into an ensemble. Caruana et al showed that ensembles perform
better when the probabilities of the base models are properly calibrated.
See their research paper "Getting the most out of ensemble selection"
(2006).

Platt scaling is easily implemented in mlr as a stacked learner using a glm
superlearner trained on cross-validated predictions of the underlying
learner. I don't think mlr can do isotonic regression currently.
